drivers/flash/flash_sam0.c

```diff
@@ -339,9 +343,8 @@ static int flash_sam0_write_protection(struct device *dev, bool enable)
 
 	flash_sam0_sem_take(dev);
 
-	for (offset = 0; offset < CONFIG_FLASH_SIZE * 1024;
-	     offset += LOCK_REGION_SIZE) {
-		*FLASH_MEM(offset) = 0U;
+	for (offset = 0; offset < CONFIG_FLASH_SIZE * 1024;  offset += LOCK_REGION_SIZE) {
+		NVMCTRL->ADDR.reg = offset;
```

Comment:
I believe this should be `NVMCTRL->ADDR.reg = offset + CONFIG_FLASH_BASE_ADDRESS;`.

As I interpret it, the whole `*FLASH_MEM(offset) = 0U;` thing that's done is an (ab)use of the note about the address register "This register is also automatically updated when writing to the page buffer.", so it's trying to do a write at the address which resolves to `offset + CONFIG_FLASH_BASE_ADDRESS`, but not actually filling/issuing the command so the only effect "should" be the address update.  It's not clear to me why this idiom was used instead of the more direct just loading the address then sending the command (like you changed this one to).

Reply:
You are of course right.
The reason this still worked was that `CONFIG_FLASH_BASE_ADDRESS` is 0 - fixed.